File: frontend/lib/battle.ts
// HU 人机对战：API 客户端 + 问题手本地存储。
// 无状态协议：/act 只需回传 {deal_seed, hero_pos, history, action, size}，服务端重放全部状态。
import { API_BASE, streamText } from "./api";

// ---------- 类型（对齐后端 engine.to_public / grade）----------
export interface BattleSizeOption {
  id: string;
  label: string;
  amount_bb: number;
}

/** 每个英雄决策的判分（翻前/翻后字段并集，按需出现）。 */
export interface DecisionGrade {
  street: string;
  spot?: string;
  spot_label: string;
  hand_class?: string;
  action: string;
  size?: string | null;
  optimal_action: string | null;
  grade: "optimal" | "acceptable" | "mistake" | "ungraded";
  correct?: boolean | null;
  severity?: number;
  ev_loss_proxy?: number;
  grounded?: boolean;
  // 翻前特有：该手牌类别在此 spot 的动作频率（GTO 范围表，权威）
  frequencies?: Record<string, number>;
  // 翻后特有
  made_label?: string;
  draw_label?: string;
  tier?: string;
  equity?: number;
  reasons?: string[];
  // 翻后建议详情（accept / 需要胜率 / MDF / 范围·坚果优势标签等）
  recommendation?: {
    recommended?: string;
    accept?: string[];
    required_equity?: number | null;
    mdf?: number | null;
    range_label?: string;
    nut_label?: string;
    size_advice?: string;
  };
}

export interface HistoryEvent {
  actor: "hero" | "villain";
  pos: string;
  street: string;
  action: string;
  size: string | null;
  amount_to: number;
  label: string;
  hero_grade?: DecisionGrade;
}

export interface BattleReview {
  is_problem: boolean;
  is_big: boolean;
  mistakes: number;
  max_severity: number;
  hero_net: number;
}

export interface BattleResult {
  reason: "fold" | "showdown";
  winner: "hero" | "villain" | "split";
  hero_net: number;
  pot_bb: number;
  villain: string[];
  villain_glyphs: string[];
  villain_class: string;
  board: string[];
  review: BattleReview;
}

export interface BattleState {
  deal_seed: number;
  matchup: string;
  matchup_label: string;
  hero_pos: string;
  villain_pos: string;
  opener_pos: string;
  defender_pos: string;
  blinds: { sb: number; bb: number };
  start_stack: number;
  street: "preflop" | "flop" | "turn" | "river";
  board: string[];
  board_glyphs: string[];
  hero: string[];
  hero_glyphs: string[];
  hero_class: string;
  pot_bb: number;
  to_call_bb: number;
  hero_stack_bb: number;
  villain_stack_bb: number;
  to_act: "hero" | null;
  complete: boolean;
  available_actions: string[];
  action_labels: Record<string, string>;
  bet_sizes: BattleSizeOption[];
  raise_sizes: BattleSizeOption[];
  history: HistoryEvent[];
  villain_last: HistoryEvent | null;
  message: string;
  grades: DecisionGrade[];
  result: BattleResult | null;
}

// ---------- API ----------
export interface BattleMatchup {
  matchup: string;
  label: string;
  opener: string;
  defender: string;
  positions: string[];
}

export async function loadBattleMatchups(): Promise<BattleMatchup[]> {
  const res = await fetch(`${API_BASE}/api/battle/matchups`, { cache: "no-store" });
  if (!res.ok) throw new Error(`battle matchups ${res.status}`);
  return (await res.json()).matchups;
}

export async function newBattle(body?: {
  matchup?: string;
  hero_pos?: string;
  seed?: number;
}): Promise<BattleState> {
  const res = await fetch(`${API_BASE}/api/battle/new`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body ?? {}),
  });
  if (!res.ok) {
    const d = await res.json().catch(() => ({}));
    throw new Error(d.detail || `battle new ${res.status}`);
  }
  return (await res.json()).state;
}

export async function battleAct(body: {
  deal_seed: number;
  matchup: string;
  hero_pos: string;
  history: HistoryEvent[];
  action: string;
  size?: string;
}): Promise<BattleState> {
  const res = await fetch(`${API_BASE}/api/battle/act`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const d = await res.json().catch(() => ({}));
    throw new Error(d.detail || `battle act ${res.status}`);
  }
  return (await res.json()).state;
}

// ---------- 对局记录（本地存储 + AI 复盘）----------
/** 一手完整对局记录（无论是否问题手都保存，便于随时问 AI）。 */
export interface RecordedHand {
  ts: number;
  hero_glyphs: string[];
  hero_pos: string;
  villain_pos: string;
  board_glyphs: string[];
  villain_glyphs: string[]; // 仅摊牌可见
  villain_class: string; // 仅摊牌可见
  hero_net: number;
  reason: string;
  winner: string;
  is_problem: boolean;
  is_big: boolean;
  max_severity: number;
  decisions: DecisionGrade[]; // 本手全部英雄决策（含最优/可接受/偏离）
  history: HistoryEvent[]; // 完整行动线（双方、分街）
}

const ACTOR_CN: Record<string, string> = { hero: "你", villain: "对手" };
const STREET_CN: Record<string, string> = {
  preflop: "翻前",
  flop: "翻牌",
  turn: "转牌",
  river: "河牌",
};

/** 把一手的完整历史整理成「分街行动线」，用于展示与喂给 AI。 */
export function actionLineFromHistory(history: HistoryEvent[]): { street: string; text: string }[] {
  const order = ["preflop", "flop", "turn", "river"];
  const groups: Record<string, string[]> = {};
  for (const e of history ?? []) {
    const who = ACTOR_CN[e.actor] ?? e.actor;
    let act = e.label ?? e.action;
    if ((e.action === "bet" || e.action === "raise") && e.amount_to) {
      act = `${act} ${e.amount_to}`;
    }
    (groups[e.street] ??= []).push(`${who}${act}`);
  }
  return order
    .filter((s) => groups[s]?.length)
    .map((s) => ({ street: STREET_CN[s] ?? s, text: groups[s].join(" · ") }));
}

export function handActionLine(h: RecordedHand): { street: string; text: string }[] {
  return actionLineFromHistory(h.history ?? []);
}

/** 兼容旧命名：分析负载复用同一结构。 */
export type ProblemHand = RecordedHand;

const KEY = "poker_battle_history_v1";
const OLD_KEY = "poker_battle_hands_v1"; // 旧版仅存问题手，加载时迁移
const MAX = 300;

function isBrowser(): boolean {
  return typeof window !== "undefined" && !!window.localStorage;
}

/** 读取全部对局记录（首次会迁移旧「问题手」存储）。 */
export function loadHands(): RecordedHand[] {
  if (!isBrowser()) return [];
  try {
    const raw = window.localStorage.getItem(KEY);
    if (raw) {
      const arr = JSON.parse(raw);
      return Array.isArray(arr) ? (arr as RecordedHand[]) : [];
    }
    const old = window.localStorage.getItem(OLD_KEY);
    if (old) {
      const arr = JSON.parse(old);
      if (Array.isArray(arr)) {
        const migrated = (arr as Record<string, unknown>[]).map((h) => ({
          villain_glyphs: [],
          villain_class: "",
          is_problem: true,
          history: [],
          ...h,
        })) as unknown as RecordedHand[];
        window.localStorage.setItem(KEY, JSON.stringify(migrated));
        return migrated;
      }
    }
    return [];
  } catch {
    return [];
  }
}

/** 合并外部（云端）记录到本地存储，按 ts 去重，返回合并后的最新列表。 */
export function mergeHands(incoming: RecordedHand[]): RecordedHand[] {
  const list = loadHands();
  const seen = new Set(list.map((h) => h.ts));
  let changed = false;
  for (const h of incoming) {
    if (h && !seen.has(h.ts)) {
      list.push(h);
      seen.add(h.ts);
      changed = true;
    }
  }
  list.sort((a, b) => a.ts - b.ts);
  const trimmed = list.length > MAX ? list.slice(list.length - MAX) : list;
  if (changed && isBrowser()) {
    try {
      window.localStorage.setItem(KEY, JSON.stringify(trimmed));
    } catch {
      /* ignore quota */
    }
  }
  return trimmed;
}

export function clearHands(): void {
  if (!isBrowser()) return;
  try {
    window.localStorage.removeItem(KEY);
    window.localStorage.removeItem(OLD_KEY);
  } catch {
    /* ignore */
  }
}

/** 一手结束后记录（每手都存），返回最新列表。 */
export function recordHand(state: BattleState): RecordedHand[] {
  const r = state.result;
  if (!r) return loadHands();
  const showdown = r.reason === "showdown";
  const hand: RecordedHand = {
    ts: Date.now(),
    hero_glyphs: state.hero_glyphs,
    hero_pos: state.hero_pos,
    villain_pos: state.villain_pos,
    board_glyphs: r.board.map(glyph),
    villain_glyphs: showdown ? r.villain.map(glyph) : [],
    villain_class: showdown ? r.villain_class : "",
    hero_net: r.hero_net,
    reason: r.reason,
    winner: r.winner,
    is_problem: r.review.is_problem,
    is_big: r.review.is_big,
    max_severity: r.review.max_severity,
    decisions: state.grades,
    history: state.history,
  };
  const list = loadHands();
  list.push(hand);
  const trimmed = list.length > MAX ? list.slice(list.length - MAX) : list;
  if (isBrowser()) {
    try {
      window.localStorage.setItem(KEY, JSON.stringify(trimmed));
    } catch {
      /* ignore quota */
    }
  }
  return trimmed;
}

const SUIT_GLYPH: Record<string, string> = { s: "♠", h: "♥", d: "♦", c: "♣" };
function glyph(card: string): string {
  return `${card[0]}${SUIT_GLYPH[card[1]?.toLowerCase()] ?? card[1] ?? ""}`;
}

function mapDecision(d: DecisionGrade) {
  return {
    street: d.street,
    spot_label: d.spot_label,
    hand_class: d.hand_class ?? "",
    action: d.action,
    optimal_action: d.optimal_action,
    grade: d.grade,
    made_label: d.made_label,
    draw_label: d.draw_label,
    tier: d.tier,
    equity: d.equity,
    reasons: d.reasons ?? [],
  };
}

function handPayload(h: RecordedHand) {
  return {
    hero_glyphs: h.hero_glyphs,
    hero_pos: h.hero_pos,
    villain_pos: h.villain_pos,
    board_glyphs: h.board_glyphs,
    villain_glyphs: h.villain_glyphs ?? [],
    villain_class: h.villain_class ?? "",
    hero_net: h.hero_net,
    reason: h.reason,
    winner: h.winner,
    action_line: handActionLine(h).map((x) => `${x.street} ${x.text}`),
    decisions: h.decisions.map(mapDecision),
  };
}

/** 批量复盘（问题手）：流式，onChunk 收到累计全文，返回最终全文。 */
export async function streamAnalyzeProblemHands(
  hands: RecordedHand[],
  onChunk: (full: string) => void,
): Promise<string> {
  return streamText(
    `${API_BASE}/api/battle/analyze/stream`,
    { hands: hands.map(handPayload) },
    onChunk,
  );
}

/** 单手复盘：问 AI「这手打得对不对、为什么」。流式返回。 */
export async function streamExplainHand(
  hand: RecordedHand,
  onChunk: (full: string) => void,
): Promise<string> {
  return streamText(
    `${API_BASE}/api/battle/explain/stream`,
    { hand: handPayload(hand) },
    onChunk,
  );
}

File: backend/app/api/battle.py
"""HU 人机对战 API：发新手 /new、行动 /act、问题手 AI 复盘 /analyze。

无状态：/act 只需 {deal_seed, hero_pos, history, action, size}，服务端重放出全部状态。
对手底牌只在摊牌时随 result 下发。判分复用翻前范围表 + 翻后启发式引擎。
"""
from __future__ import annotations

import random
from typing import Dict, List, Optional

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from app.llm import get_provider
from app.poker.battle import engine as E

_STREAM_HEADERS = {"X-Accel-Buffering": "no", "Cache-Control": "no-cache"}

router = APIRouter(prefix="/battle", tags=["battle"])


@router.get("/matchups")
def battle_matchups() -> dict:
    """可选对位（不同位置对抗）。每个对位可选英雄扮演开池方或防守方。"""
    from app.poker.battle.deal import MATCHUPS, matchup_label, matchup_positions

    out = []
    for m in MATCHUPS:
        opener, defender = matchup_positions(m)
        out.append(
            {
                "matchup": m,
                "label": matchup_label(m),
                "opener": opener,
                "defender": defender,
                "positions": [opener, defender],
            }
        )
    return {"matchups": out}


class NewRequest(BaseModel):
    matchup: Optional[str] = Field(None, description="对位 vs_RFI 键，如 BB_vs_BTN；缺省随机")
    hero_pos: Optional[str] = Field(None, description="英雄位置（须属于该对位）；缺省随机取一侧")
    seed: Optional[int] = Field(None, description="发牌种子，缺省随机（可复现用）")


def _resolve_matchup_hero(matchup: Optional[str], hero_pos: Optional[str]) -> tuple:
    """解析对位 + 英雄位置。兼容旧的仅传 hero_pos=BTN/BB（默认对位 BB_vs_BTN）。"""
    from app.poker.battle.deal import MATCHUPS, matchup_positions

    if matchup is None:
        if hero_pos in ("BTN", "BB") or hero_pos is None:
            matchup = "BB_vs_BTN"  # 向后兼容默认对位
        else:
            raise HTTPException(status_code=400, detail="缺少 matchup，且 hero_pos 非 BTN/BB")
    if matchup not in MATCHUPS:
        raise HTTPException(status_code=400, detail=f"对位不存在：{matchup}")
    opener, defender = matchup_positions(matchup)
    if hero_pos is None:
        hero_pos = random.choice([opener, defender])
    if hero_pos not in (opener, defender):
        raise HTTPException(
            status_code=400, detail=f"hero_pos {hero_pos} 不属于对位 {matchup}（{opener}/{defender}）"
        )
    return matchup, hero_pos


@router.post("/new")
def battle_new(req: NewRequest) -> dict:
    matchup, hero_pos = _resolve_matchup_hero(req.matchup, req.hero_pos)
    seed = req.seed if req.seed is not None else random.getrandbits(32)
    try:
        b = E.new_hand(seed, matchup, hero_pos)
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(status_code=500, detail=f"发牌失败：{exc}") from exc
    return {"state": b.to_public()}


class ActRequest(BaseModel):
    deal_seed: int
    matchup: Optional[str] = Field(None, description="对位 vs_RFI 键；缺省按 hero_pos 兼容推断")
    hero_pos: str = Field(..., description="英雄位置")
    history: List[Dict] = Field(default_factory=list)
    action: str = Field(..., description="fold/check/call/bet/raise")
    size: Optional[str] = Field(None, description="下注/加注尺度 id")


@router.post("/act")
def battle_act(req: ActRequest) -> dict:
    matchup, hero_pos = _resolve_matchup_hero(req.matchup, req.hero_pos)
    try:
        b = E.act(req.deal_seed, matchup, hero_pos, req.history, req.action, req.size)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(status_code=500, detail=f"对战推进失败：{exc}") from exc
    return {"state": b.to_public()}


# ---------------- 问题手 AI 复盘 ----------------
class HandDecision(BaseModel):
    street: str = ""
    spot_label: str = ""
    hand_class: str = ""
    action: str = ""
    optimal_action: Optional[str] = None
    grade: str = ""
    made_label: Optional[str] = None
    draw_label: Optional[str] = None
    tier: Optional[str] = None
    equity: Optional[float] = None
    reasons: List[str] = Field(default_factory=list)


class ProblemHand(BaseModel):
    hero_glyphs: List[str] = Field(default_factory=list)
    hero_pos: str = ""
    villain_pos: str = ""
    board_glyphs: List[str] = Field(default_factory=list)
    villain_glyphs: List[str] = Field(default_factory=list)  # 仅摊牌
    villain_class: str = ""  # 仅摊牌
    hero_net: Optional[float] = None
    reason: str = ""  # fold / showdown
    winner: str = ""
    action_line: List[str] = Field(default_factory=list)  # 分街行动线（可读）
    decisions: List[HandDecision] = Field(default_factory=list)


class BattleAnalyzeRequest(BaseModel):
    hands: List[ProblemHand] = Field(default_factory=list)


class BattleExplainRequest(BaseModel):
    hand: ProblemHand


ACTION_CN = {"fold": "弃牌", "call": "跟注", "check": "过牌", "bet": "下注", "raise": "加注"}


def _act_cn(a: Optional[str]) -> str:
    return ACTION_CN.get(a or "", a or "-")


BATTLE_REVIEW_SYSTEM = (
    "你是一名职业德州扑克教练，正在复盘学员和 AI 对战中被系统标记为「有问题」的手牌。"
    "每手牌的判分（最优动作、评级、牌力/胜率、牌面纹理与依据）都是引擎给出的**权威事实**，"
    "你只能基于它们分析，绝不能编造未提供的数字或牌。用简洁中文、分点作答，务实、可执行。"
)


def build_battle_prompt(hands: List[ProblemHand]) -> str:
    lines: List[str] = [f"共 {len(hands)} 手被标记的问题手，逐手事实如下：\n"]
    for i, h in enumerate(hands, 1):
        board = " ".join(h.board_glyphs) or "（未到翻牌/未摊牌）"
        hero = " ".join(h.hero_glyphs)
        net = f"{h.hero_net:+.1f}bb" if h.hero_net is not None else "?"
        lines.append(f"[手 {i}] {h.hero_pos} 持 {hero}，牌面 {board}，结果 {net}（{h.winner or '-'}）")
        if h.action_line:
            lines.append("   行动线：" + " ｜ ".join(h.action_line))
        for d in h.decisions:
            if d.grade not in ("mistake", "acceptable"):
                continue
            hand_desc = d.made_label or ""
            if d.draw_label:
                hand_desc += f"+{d.draw_label}"
            eq = f"，胜率 {d.equity:.0%}" if d.equity is not None else ""
            opt = f"，应{_act_cn(d.optimal_action)}" if d.optimal_action else ""
            why = ("；依据：" + "；".join(d.reasons)) if d.reasons else ""
            flag = "❌偏离" if d.grade == "mistake" else "⚠可接受但非首选"
            lines.append(
                f"   - {d.spot_label}·{d.hand_class}（{hand_desc or '—'}{eq}）："
                f"选了{_act_cn(d.action)}{opt} [{flag}]{why}"
            )
        lines.append("")
    lines.append(
        "请基于以上事实复盘，中文输出：\n"
        "1) 总体倾向：1-2 句点出最突出的漏洞（过紧/过松/太被动/太激进/尺度/线路）；\n"
        "2) 主要漏洞：2-4 条，按严重度排序，每条引用上面具体的手牌/牌面/位置并说明为何是漏洞；\n"
        "3) 修正建议：每条可直接执行；\n"
        "4) 下一步训练重点：具体到翻前/翻后与位置。\n"
        "只依据上面提供的事实，不要编造未出现的牌或数字。"
    )
    return "\n".join(lines)


@router.post("/analyze")
def battle_analyze(req: BattleAnalyzeRequest) -> dict:
    if not req.hands:
        raise HTTPException(status_code=400, detail="暂无问题手可分析，先去对战积累几手吧")

    provider = get_provider()
    if not (provider.gateway_ready or provider.openai_ready):
        raise HTTPException(status_code=503, detail="LLM 未配置（见 backend/.env.example）")

    prompt = build_battle_prompt(req.hands)
    try:
        text = provider.text(prompt, system=BATTLE_REVIEW_SYSTEM, max_tokens=900, model="gpt-4o")
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(status_code=502, detail=f"LLM 调用失败：{exc}") from exc
    if not text or not text.strip():
        raise HTTPException(status_code=502, detail="复盘生成为空，请重试")

    return {"report": text, "analyzed": len(req.hands)}


@router.post("/analyze/stream")
def battle_analyze_stream(req: BattleAnalyzeRequest) -> StreamingResponse:
    """与 /battle/analyze 相同，但纯文本流式返回，边收边渲染。"""
    if not req.hands:
        raise HTTPException(status_code=400, detail="暂无问题手可分析，先去对战积累几手吧")

    provider = get_provider()
    if not (provider.gateway_ready or provider.openai_ready):
        raise HTTPException(status_code=503, detail="LLM 未配置（见 backend/.env.example）")

    prompt = build_battle_prompt(req.hands)

    def gen():
        # 用非推理的 gpt-4o：逐字即时流出，避免推理模型"先思考后爆发"导致流式体感无提升。
        yield from provider.text_stream(prompt, system=BATTLE_REVIEW_SYSTEM, max_tokens=900, model="gpt-4o")

    return StreamingResponse(gen(), media_type="text/plain; charset=utf-8", headers=_STREAM_HEADERS)


# ---------------- 单手复盘：这手打得对不对 ----------------
GRADE_CN = {"optimal": "最优", "acceptable": "可接受", "mistake": "偏离", "ungraded": "未判分"}


def build_explain_prompt(h: ProblemHand) -> str:
    board = " ".join(h.board_glyphs) or "（未到翻牌/未摊牌）"
    hero = " ".join(h.hero_glyphs)
    net = f"{h.hero_net:+.1f}bb" if h.hero_net is not None else "?"
    villain = " ".join(h.villain_glyphs)
    if h.reason == "showdown" and villain:
        villain_line = f"对手 {h.villain_pos} 摊牌 {villain}" + (f"（{h.villain_class}）" if h.villain_class else "")
    elif h.reason == "fold":
        villain_line = f"对手 {h.villain_pos} 未摊牌（有人弃牌成交）"
    else:
        villain_line = f"对手 {h.villain_pos}（底牌未知）"

    lines: List[str] = [
        "复盘这一手单挑（HU，100bb）的英雄决策是否正确：",
        f"英雄 {h.hero_pos} 持 {hero}；{villain_line}",
        f"牌面 {board}；本手英雄净收益 {net}（{h.winner or '-'}）",
    ]
    if h.action_line:
        lines.append("行动线：" + " ｜ ".join(h.action_line))
    lines += [
        "",
        "本手英雄逐个决策（含引擎判分，判分为权威事实）：",
    ]
    if not h.decisions:
        lines.append("（本手英雄无需决策或未记录）")
    for d in h.decisions:
        hand_desc = d.made_label or d.hand_class or ""
        if d.draw_label:
            hand_desc += f"+{d.draw_label}"
        eq = f"，胜率 {d.equity:.0%}" if d.equity is not None else ""
        opt = f"，引擎建议{_act_cn(d.optimal_action)}" if d.optimal_action else ""
        why = ("；依据：" + "；".join(d.reasons)) if d.reasons else ""
        tag = GRADE_CN.get(d.grade, d.grade)
        lines.append(
            f"   - {d.spot_label}·{d.hand_class}（{hand_desc or '—'}{eq}）："
            f"实际{_act_cn(d.action)}{opt} [{tag}]{why}"
        )
    lines.append("")
    lines.append(
        "请用简洁中文回答，务实、可执行：\n"
        "1) 结论：这手整体打得对不对（一句话，含总体评价）；\n"
        "2) 逐街点评：翻前/翻牌/转牌/河牌各自的决策是否正确、为什么（引用上面的胜率/牌力/建议）；\n"
        "3) 更优线路：若有更好的打法请具体说明尺度与理由；没有则说明为何当前已足够好。\n"
        "只依据上面提供的事实，不要编造未出现的牌或数字。若对手未摊牌，不要臆测其底牌。"
    )
    return "\n".join(lines)


EXPLAIN_SYSTEM = (
    "你是一名职业德州扑克教练，正在针对学员和 AI 单挑中的**某一手牌**做逐街复盘。"
    "每个决策的判分（最优动作、评级、牌力/胜率、依据）都是引擎给出的**权威事实**，"
    "你只能据此分析，绝不能编造未提供的数字或牌。用简洁中文、分点作答。"
)


@router.post("/explain")
def battle_explain(req: BattleExplainRequest) -> dict:
    provider = get_provider()
    if not (provider.gateway_ready or provider.openai_ready):
        raise HTTPException(status_code=503, detail="LLM 未配置（见 backend/.env.example）")
    prompt = build_explain_prompt(req.hand)
    try:
        text = provider.text(prompt, system=EXPLAIN_SYSTEM, max_tokens=700, model="gpt-4o")
    except Exception as exc:  # noqa: BLE001
        raise HTTPException(status_code=502, detail=f"LLM 调用失败：{exc}") from exc
    if not text or not text.strip():
        raise HTTPException(status_code=502, detail="复盘生成为空，请重试")
    return {"report": text}


@router.post("/explain/stream")
def battle_explain_stream(req: BattleExplainRequest) -> StreamingResponse:
    """单手复盘，纯文本流式返回。"""
    provider = get_provider()
    if not (provider.gateway_ready or provider.openai_ready):
        raise HTTPException(status_code=503, detail="LLM 未配置（见 backend/.env.example）")
    prompt = build_explain_prompt(req.hand)

    def gen():
        yield from provider.text_stream(prompt, system=EXPLAIN_SYSTEM, max_tokens=700, model="gpt-4o")

    return StreamingResponse(gen(), media_type="text/plain; charset=utf-8", headers=_STREAM_HEADERS)
